core/server/api_container/server/docker_compose_transpiler/docker_compose_transpiler.go

```diff
@@ -0,0 +1,582 @@
+	}
+	return imageBuildSpecKurtosisType, nil
+}
+
+// TODO: Support public ports
+func getStarlarkPortSpecs(composePorts []types.ServicePortConfig) (*starlark.Dict, error) {
+	portSpecs := starlark.NewDict(len(composePorts))
+
+	for portIdx, dockerPort := range composePorts {
+		portName := fmt.Sprintf("port%d", portIdx)
+
+		dockerProto := dockerPort.Protocol
+		kurtosisProto, found := dockerPortProtosToKurtosisPortProtos[strings.ToLower(dockerProto)]
+		if !found {
+			return nil, stacktrace.NewError("Port #%d has unsupported protocol '%v'", portIdx, dockerProto)
+		}
+
+		portSpec, interpretationErr := port_spec_starlark.CreatePortSpecUsingGoValues(
+			uint16(dockerPort.Target),
+			kurtosisProto,
+			nil, // Application protocol (which Compose doesn't have)
+			"",  // Wait timeout (which Compose doesn't have a way to override)
+		)
+		if interpretationErr != nil {
+			return nil, stacktrace.Propagate(interpretationErr, "An error occurred creating a %s object from port #%d", port_spec_starlark.PortSpecTypeName, portIdx)
+		}
+		if err := portSpecs.SetKey(starlark.String(portName), portSpec); err != nil {
+			return nil, stacktrace.Propagate(err, "An error occurred putting port #%d in Starlark dict", portIdx)
+		}
+	}
+
+	return portSpecs, nil
+}
+
+func getStarlarkEntrypoint(composeEntrypoint types.ShellCommand) *starlark.List {
+	entrypointSLStrs := make([]starlark.Value, len(composeEntrypoint))
+	for idx, entrypointFragment := range composeEntrypoint {
+		entrypointSLStrs[idx] = starlark.String(entrypointFragment)
+	}
+	return starlark.NewList(entrypointSLStrs)
+}
+
+func getStarlarkCommand(composeCommand types.ShellCommand) *starlark.List {
+	commandSLStrs := make([]starlark.Value, len(composeCommand))
+	for idx, commandFragment := range composeCommand {
+		commandSLStrs[idx] = starlark.String(commandFragment)
+	}
+	return starlark.NewList(commandSLStrs)
+}
+
+func getStarlarkEnvVars(composeEnvironment types.MappingWithEquals) (*starlark.Dict, error) {
+	// make iteration order of [composeEnvironment] deterministic by getting the keys and sorting them
+	envVarKeys := []string{}
+	for key := range composeEnvironment {
+		envVarKeys = append(envVarKeys, key)
+	}
+	sort.Strings(envVarKeys)
+	enVarsSLDict := starlark.NewDict(len(composeEnvironment))
+	for _, key := range envVarKeys {
+		value := composeEnvironment[key]
+		if value == nil {
+			continue
+		}
+		if err := enVarsSLDict.SetKey(
+			starlark.String(key),
+			starlark.String(*value),
+		); err != nil {
+			return nil, stacktrace.Propagate(err, "An error occurred setting key '%s' in environment variables Starlark dict.", key)
+		}
+	}
+	return enVarsSLDict, nil
+}
+
+// The 'volumes:' compose key supports named volumes and bind mounts https://docs.docker.com/storage/volumes/
+// bind mount semantics for starlark:
+// <rel path on host>:<path on container> := upload a files artifacts of <rel path on host>, mount the files artifacts on the container at <path on container>
```

Comment:
These kind of encoding seems a bit trick and can potentially lead to some confusion with user. Wondering if we should simplify the conversion to a safer transcription and let the user adjust the richer options in Starlark later.

Reply:
Definitely. Particularly with volume's, it's not a 1:1 match in terms of semantics, but rather it requires translating Compose's volume concept into our concepts which definitely needs to be transparent to users. 

Going to put some thought into this, but do you have any ideas for what a safer transcription might be?

Reply:
After putting some thought into it, going to merge this for now and return to this when writing docs on the translation between files artifact and volume.